firestore: query watch samples

AlisskaPie opened the pull request:
Samples:
- [x] Listen to a document 
- [x] Listen to multiple documents in a collection
- [x] View changes between snapshots
- [x] Detach a listener
- [x] Handle listen errors

AlisskaPie committed:
query watch samples

AlisskaPie committed:
listen to a snapshot

AlisskaPie committed:
add all listen samples

IlyaFaer commented on firestore/firestore_snippets/listen_changes.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+// listenChanges listens to a query, returning the list of DocumentChange
+// in the first snapshot.
```
What do you mean, the first snapshot? It's the snapshot with new changes, so it's not the first of something in my vision

AlisskaPie replied:
When we're iterating through `iter.Next()` just once, we get a first snapshot.
Also, I copied this comment from Java samples, they're similar.
[listenForChanges](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/fe0d5cb7214680461928cac1d93d1711842884b9/firestore/src/main/java/com/example/firestore/snippets/ListenDataSnippets.java#L124)

AlisskaPie replied:
How do you suggest to rewrite this comment?

IlyaFaer replied:
If it's taken from Java then let's keep it

AlisskaPie replied:
In Java they have: Listen to a query, returning the list of DocumentChange **events** in the first snapshot.
They have event methods and we don't so I decided to delete **events**.

IlyaFaer commented on firestore/firestore_snippets/listen_test.go:
```diff
@@ -0,0 +1,118 @@
+	}
+	defer client.Close()
+
+	// Set prepareQuery for all listen tests.
```
Not sure what does this comment mean

AlisskaPie replied:
Once we run this test, we'll have data in base, that can be used for other tests.
I saw something similar in Storage samples. There we need to run create bucket test first before run other tests.
I could remove this comment, I was not sure about adding it

AlisskaPie replied:
Removed

IlyaFaer commented on firestore/firestore_snippets/listen_test.go:
```diff
@@ -0,0 +1,118 @@
+	if err != nil {
+		t.Fatalf("firestore.NewClient: %v", err)
+	}
+	defer client.Close()
```
Project and client creation could be moved into separate function

AlisskaPie replied:
I thought about it, but I don't have idea how to skip test in a separate function, so I left it as it's.
Do you have an idea?

AlisskaPie replied:
Done

IlyaFaer commented on firestore/firestore_snippets/listen_changes.go:
```diff
@@ -0,0 +1,58 @@
+	}
+	defer client.Close()
+
+	snap, err := client.Collection("cities").Where("state", "==", "CA").Snapshots(ctx).Next()
```
`Next()` probably should be called within `for`. For now it'll only listen for a single one snapshot, but I suspect the samples intended to show the endless listening

AlisskaPie replied:
I watched the same samples on other languages, and it looks like they use the same way.
I tried to listen with for loop but, as I wrote you personally, I got freezes and falls for local test.

IlyaFaer replied:
Looking at Python sample I'd say it registers a callback function, so on every new snapshot it'll call the callback func (cycle). In Go there is no registering, but moving the `Next()` call into `for` cycle will do the same thing. Seems to me all the `Next()` calls must be moved into `for` cycles, or it'll be listening just for a single one change

What are the local tests failures?

IlyaFaer commented on firestore/firestore_snippets/listen_document.go:
```diff
@@ -0,0 +1,54 @@
+			return fmt.Errorf("listen failed: %v", err)
+		}
+		fmt.Fprintf(w, "Received document snapshot: %v\n", snap.Data())
+		return nil
```
`Next()` seems to be blocking, so for single snapshot we can just call `Next()`, without `for` cycle, like in `listen_errors`

AlisskaPie replied:
What do you mean by `Next()` is blocking?
Ok, I'll fix it.

AlisskaPie replied:
Done

AlisskaPie commented on firestore/firestore_snippets/listen_errors.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+// listenErrors demonstrates how to handle listening errors.
+func listenErrors(w io.Writer, projectID string) error {
```
I'm not sure about this test, but I did the same as was done in Java

AlisskaPie committed:
remove comment, remove cycle for, setup func

AlisskaPie committed:
fix comment

AlisskaPie committed:
Merge branch 'master' into query_watch_samples

AlisskaPie committed:
add a point

AlisskaPie committed:
Merge branch 'master' of https://github.com/q-logic/golang-samples into query_watch_samples

AlisskaPie committed:
Merge branch 'query_watch_samples' of https://github.com/q-logic/golang-samples into query_watch_samples

AlisskaPie committed:
t in setup, created query in listen_test.go

AlisskaPie committed:
replaced context, iterator, added loops

AlisskaPie committed:
check DeadlineExceeded, != nil, delete stop sample

AlisskaPie committed:
Update firestore/firestore_snippets/listen_test.go

Co-authored-by: Tyler Bui-Palsulich <26876514+tbpg@users.noreply.github.com>

AlisskaPie committed:
ctx passed to function, added duration for tests

AlisskaPie committed:
added if snap != nil, changes in parallel, cleanup

AlisskaPie committed:
add region tag in listenDocument for detach sample

tritone committed:
Merge branch 'master' into query_watch_samples

AlisskaPie committed:
passing buf, get previous listenChanges version

AlisskaPie committed:
Merge branch 'query_watch_samples' of https://github.com/q-logic/golang-samples into query_watch_samples

AlisskaPie committed:
Update firestore/firestore_snippets/listen_test.go

Co-authored-by: Tyler Bui-Palsulich <26876514+tbpg@users.noreply.github.com>

AlisskaPie committed:
review updates

AlisskaPie committed:
Merge branch 'query_watch_samples' of https://github.com/q-logic/golang-samples into query_watch_samples

renovate-bot committed:
chore(all): update all (#1694)

This PR contains the following updates:

| Package | Type | Update | Change |
|---|---|---|---|
| [github.com/GoogleCloudPlatform/golang-samples](https://togithub.com/GoogleCloudPlatform/golang-samples) | require | digest | `04937a2` -> `6aca2ff` |
| [github.com/GoogleCloudPlatform/golang-samples](https://togithub.com/GoogleCloudPlatform/golang-samples) | require | digest | `b993d02` -> `6aca2ff` |
| [github.com/GoogleCloudPlatform/golang-samples/run/grpc-ping](https://togithub.com/GoogleCloudPlatform/golang-samples) | require | digest | `04937a2` -> `6aca2ff` |
| [github.com/GoogleCloudPlatform/golang-samples/run/grpc-server-streaming](https://togithub.com/GoogleCloudPlatform/golang-samples) | require | digest | `04937a2` -> `6aca2ff` |
| [github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/trace](https://togithub.com/GoogleCloudPlatform/opentelemetry-operations-go) | require | minor | `v0.10.1-0.20200828070728-ba72e89a4087` -> `v0.11.0` |
| [github.com/aws/aws-sdk-go](https://togithub.com/aws/aws-sdk-go) | require | patch | `v1.34.14` -> `v1.34.18` |
| [golang.org/x/exp](https://togithub.com/golang/exp) | require | digest | `1ff542f` -> `c4f52b2` |
| [golang.org/x/net](https://togithub.com/golang/net) | require | digest | `c890458` -> `62affa3` |
| [golang.org/x/oauth2](https://togithub.com/golang/oauth2) | require | digest | `bf48bf1` -> `5d25da1` |
| [golang.org/x/tools](https://togithub.com/golang/tools) | require | digest | `5a2aa26` -> `39188db` |
| [google.golang.org/genproto](https://togithub.com/googleapis/go-genproto) | require | digest | `d751682` -> `0bd0a95` |

---

### Release Notes

<details>
<summary>aws/aws-sdk-go</summary>

### [`v1.34.18`](https://togithub.com/aws/aws-sdk-go/blob/master/CHANGELOG.md#Release-v13418-2020-09-04)

[Compare Source](https://togithub.com/aws/aws-sdk-go/compare/v1.34.17...v1.34.18)

===

##### Service Client Updates

-   `service/ssm`: Updates service documentation
    -   Documentation-only updates for AWS Systems Manager
-   `service/workspaces`: Updates service API and documentation
    -   Adding support for Microsoft Office 2016 and Microsoft Office 2019 in BYOL Images
-   `service/xray`: Updates service API and documentation
    -   Enhancing CreateGroup, UpdateGroup, GetGroup and GetGroups APIs to support configuring X-Ray Insights

### [`v1.34.17`](https://togithub.com/aws/aws-sdk-go/blob/master/CHANGELOG.md#Release-v13417-2020-09-03)

[Compare Source](https://togithub.com/aws/aws-sdk-go/compare/v1.34.16...v1.34.17)

===

##### Service Client Updates

-   `service/guardduty`: Updates service API and documentation
    -   GuardDuty findings triggered by failed events now include the error code name within the AwsApiCallAction section.
-   `service/kendra`: Updates service API and documentation
    -   Amazon Kendra now returns confidence scores for both 'answer' and 'question and answer' query responses.
-   `service/mediapackage`: Updates service API and documentation
    -   Enables inserting a UTCTiming XML tag in the output manifest of a DASH endpoint which a media player will use to help with time synchronization.
-   `service/states`: Updates service API and documentation
    -   This release of the AWS Step Functions SDK introduces support for payloads up to 256KB for Standard and Express workflows

### [`v1.34.16`](https://togithub.com/aws/aws-sdk-go/blob/master/CHANGELOG.md#Release-v13416-2020-09-02)

[Compare Source](https://togithub.com/aws/aws-sdk-go/compare/v1.34.15...v1.34.16)

===

##### Service Client Updates

-   `service/ec2`: Updates service API and documentation
    -   This release adds a new transit gateway attachment state and resource type.
-   `service/macie2`: Updates service API and documentation

### [`v1.34.15`](https://togithub.com/aws/aws-sdk-go/blob/master/CHANGELOG.md#Release-v13415-2020-09-01)

[Compare Source](https://togithub.com/aws/aws-sdk-go/compare/v1.34.14...v1.34.15)

===

##### Service Client Updates

-   `service/codeguru-reviewer`: Updates service API and documentation
-   `service/securityhub`: Updates service API and documentation

</details>

---

### Renovate configuration

:date: **Schedule**: "before 3am on Monday" (UTC).

:vertical_traffic_light: **Automerge**: Disabled by config. Please merge this manually once you are satisfied.

:recycle: **Rebasing**: Renovate will not automatically rebase this PR, because other commits have been found.

:ghost: **Immortal**: This PR will be recreated if closed unmerged. Get [config help](https://togithub.com/renovatebot/config-help/issues) if that's undesired.

---

 - [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check this box

---

This PR has been generated by [WhiteSource Renovate](https://renovate.whitesourcesoftware.com). View repository job log [here](https://app.renovatebot.com/dashboard#github/GoogleCloudPlatform/golang-samples).

shubha-rajan committed:
docs(cloudsql): add Postgres GAE Flex sample comment (#1693)

Added comment in Cloud SQL Postgres sample to explain workaround for App Engine Flex Go runtime, where Go modules are not supported.

tritone committed:
feat(storage): add CustomTime to metadata sample (#1692)

This also requires bumping the dependency on storage to
capture this newly added field.

tbpg committed:
Merge branch 'master' into query_watch_samples
